### PR TITLE
Add a post on how the composable matchers work.

### PR DESCRIPTION
**Pull request by samphippen**

### CODE REVIEW

**Comment by xaviershay** [comment]
Overall content is good, I'm leaving comments on style.

File: source/blog/2015-05-08-rspec-compsable-matchers.md [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+
+The new RSpec matchers make heavy use of `===` internally, if you'd like to get
+more of an understanding of `===`, I'd heavily suggest you pop open IRB and
+play around with it a little before moving on.
+
+## RSpec matchers
+
+In an RSpec expectation expression like `expect(object_1).to eq(object_2)`
+there are two primary objects at play, the **target** and the **matcher**. The
+**target** object is created by "expect" and basically wraps the passed object
```
Comment (xaviershay) [inline source/blog/2015-05-08-rspec-compsable-matchers.md]: remove all occurrences of "basically", it doesn't add anything.

File: source/blog/2015-05-08-rspec-compsable-matchers.md [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+to it in a class called `ExpectationTarget`. The primary methods on the
+`ExpectationTarget` that users see are `to` and `not_to`/`to_not`. Let's focus
+on `to`.
+
+The `to` method is basically responsible for invoking the `matches?` method on the
+passed matcher. It's possible to see this happening quite easily in an IRB session.
```
Comment (xaviershay) [inline source/blog/2015-05-08-rspec-compsable-matchers.md]: "quite easily" is subjective and adds nothing to this sentence.

File: source/blog/2015-05-08-rspec-compsable-matchers.md [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+a very simple definition of `match`. For example, here's the definition of `match`
+for the `eq` matcher:
+
+```ruby
+def match(expected, actual)
+  actual == expected
+end
+```
+
+Some matchers, however, have a more complicated match protocol. This more complicated
```
Comment (xaviershay) [inline source/blog/2015-05-08-rspec-compsable-matchers.md]: Move however to beginning or end, or remove it. No need for it in the middle with two commas.

File: source/blog/2015-05-08-rspec-compsable-matchers.md [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+protocol is called the **composable** matcher protocol, which is what we'll discuss next.
+
+## Composable RSpec matchers
+
+A composable RSpec matcher is one which expects other RSpec matchers for some/all of it's
+arguments. The `match` matcher is perhaps an obvious place to start. In RSpec 2.xx the match
```
Comment (xaviershay) [inline source/blog/2015-05-08-rspec-compsable-matchers.md]: remove all occurrences of "obvious". Obvious to you maybe, not your reader.

File: source/blog/2015-05-08-rspec-compsable-matchers.md [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+matcher was really only used match regexes, and its match definition looked thus:
```
Comment (xaviershay) [inline source/blog/2015-05-08-rspec-compsable-matchers.md]: is "really only" necessary here?

File: source/blog/2015-05-08-rspec-compsable-matchers.md [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+
+```ruby
+def match(expected, actual)
+  actual.match expected
+end
+```
+
+Now obviously, one could extend the capabilites by providing other objects that provided a
+definition of match, but that's besides the point.
```
Comment (xaviershay) [inline source/blog/2015-05-08-rspec-compsable-matchers.md]: This paragraph neither obvious nor necessary.

File: source/blog/2015-05-08-rspec-compsable-matchers.md [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+
+The new definition of the match matcher's match method is thus:
```
Comment (xaviershay) [inline source/blog/2015-05-08-rspec-compsable-matchers.md]: "match matcher's match" try again ;)

File: source/blog/2015-05-08-rspec-compsable-matchers.md [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+
+```ruby
+def match(expected, actual)
+  return true if values_match?(expected, actual)
+  return false unless can_safely_call_match?(expected, actual)
+  actual.match(expected)
+end
+```
+
+The key important difference is that call to `values_match?` which is the core implementation
```
Comment (xaviershay) [inline source/blog/2015-05-08-rspec-compsable-matchers.md]: "key" is redundant in combination with "important"

File: source/blog/2015-05-08-rspec-compsable-matchers.md [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+end
+```
+
+The first thing to note is the special casing for Hashes and Arrays. Basically,
+what those methods do is recursively apply `values_match?` to each of the
+elements in the various collection types as you'd expect. Arrays by scanning
+elements and hashes by matching keys and values.
+
+The next thing the method does is return `true` if expected === actual. An easy
+way to demonstrate this is back in our IRB session:
```
Comment (xaviershay) [inline source/blog/2015-05-08-rspec-compsable-matchers.md]: Not easy, it's totally weird.

File: source/blog/2015-05-08-rspec-compsable-matchers.md [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+  matches?(value)
+end
+```
+
+This means that `expected` in the expression `expected === actual` can be another RSpec matcher,
+a regex, or any other object that responds to `===`. This means, for example, that one can trivially
+match a lambda by doing `expect(obj).to match(lambda { |actual| ... })`. Given that all RSpec matchers
+implement `===` we are able to compose them arbitrarily with other "matching" values in the Ruby system.
+
+An example I've given in the past is matching nesting hashes like so:
```
Comment (xaviershay) [inline source/blog/2015-05-08-rspec-compsable-matchers.md]: "given in the past" is useless information.

File: source/blog/2015-05-08-rspec-compsable-matchers.md [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+
+```
+expect({:data => {... complex hash ...}).to include(:data => a_hash_including(:response => "success"))
+```
+
+The reason we're able to pass the `a_hash_including` matcher inside the
+`include` matcher is that `values_match?` will dilligently call `===` on the
+inner, `a_hash_including` matcher, which will invoke its `matches?` method, which
```
Comment (xaviershay) [inline source/blog/2015-05-08-rspec-compsable-matchers.md]: comma after "inner"?

File: source/blog/2015-05-08-rspec-compsable-matchers.md [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+implement `===` we are able to compose them arbitrarily with other "matching" values in the Ruby system.
+
+An example I've given in the past is matching nesting hashes like so:
+
+```
+expect({:data => {... complex hash ...}).to include(:data => a_hash_including(:response => "success"))
+```
+
+The reason we're able to pass the `a_hash_including` matcher inside the
+`include` matcher is that `values_match?` will dilligently call `===` on the
```
Comment (xaviershay) [inline source/blog/2015-05-08-rspec-compsable-matchers.md]: what does "dilligently" mean here? "always"? Is it redundant?

File: source/blog/2015-05-08-rspec-compsable-matchers.md [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+inner, `a_hash_including` matcher, which will invoke its `matches?` method, which
+will allow it to match the complex inner hash.
+
+
+##Conclusions
+
+The old RSpec match protocol used to be a lot easier to understand. Each of the
+matchers held their own matching logic, which was universally pretty simple. It
```
Comment (xaviershay) [inline source/blog/2015-05-08-rspec-compsable-matchers.md]: "pretty simple"

File: source/blog/2015-05-08-rspec-compsable-matchers.md [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+did mean, however, that one could not use RSpec matchers in a highly composable
```
Comment (xaviershay) [inline source/blog/2015-05-08-rspec-compsable-matchers.md]: "You couldn't use them in a composable way though."

File: source/blog/2015-05-08-rspec-compsable-matchers.md [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+way. For example: performing nested inclusion matches would require multiple
+expectation expressions.
```
Comment (xaviershay) [inline source/blog/2015-05-08-rspec-compsable-matchers.md]: maybe a code example/comparison makes sense here?

File: source/blog/2015-05-08-rspec-compsable-matchers.md [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+
+I think the new RSpec match protocol is a lot "simpler". There is one core
```
Comment (xaviershay) [inline source/blog/2015-05-08-rspec-compsable-matchers.md]: What does '"simpler"' mean? Why the scare quotes?

File: source/blog/2015-05-08-rspec-compsable-matchers.md [hunk 17]
```diff
@@ -0,0 +1,216 @@
...
+piece of matching logic, the `values_match?` method. Once you understand that
+this method is at the core of all matches, it becomes a lot easier to reason
+about what's going on.  All matches now look like a 'tree'. At the root of the
+tree we use the `matches?` method to descend through a potentially complex tree
```
Comment (xaviershay) [inline source/blog/2015-05-08-rspec-compsable-matchers.md]: "complex branches" maybe? You said tree twice already.

**Comment by xaviershay** [comment]
tl;dr see how many words you can remove from each sentence without changing its meaning :)

File: source/blog/2015-05-08-rspec-compsable-matchers.md [hunk 18]
```diff
@@ -0,0 +1,216 @@
...
+a subclass of array. For regexes, `===` returns true if the passed string
+matches the regex, and false if it does not.
+
+The `===` method is used in Ruby case statements to decide if a branch should
+be taken. For example
+
+```
+case 3
+when (1...18)
+    puts "hello"
```
Comment (JonRowe) [inline source/blog/2015-05-08-rspec-compsable-matchers.md]: Extremly minor quibble but this should probably only be indented two spaces?

**Comment by JonRowe** [comment]
Pretty much :+1: to all @xaviershay's comments, its pretty much a writing style thing tho

**Comment by myronmarston** [comment]
Thanks for putting this together, @samphippen.  It looks like your local branch wasn't very current when you committed this so this branch is missing lots of improvements from master.  Mind rebasing against `source`?  It'll help me be able to view the blog post on a local server to see how it renders (and that's probably something you should check, too).

Also, our markdown engine uses `~~~ ruby` instead of triple backticks for fencing code blocks so they are rendering funny currently.  Please update that as well.

I'll go review the content now.

File: source/blog/2015-05-08-rspec-compsable-matchers.md [hunk 19]
```diff
@@ -0,0 +1,216 @@
...
+
+In Ruby, a number of objects implement the `===` protocol. Some of the objects
+that implement this protocol include:
+
+* all class objects
+* lambdas
+* regexes
+* ranges
+
+Conceptually, this protocol represents a "match". That is, the object that
```
Comment (myronmarston) [inline source/blog/2015-05-08-rspec-compsable-matchers.md]: I've also found it useful to explain this protocol in terms of it defining a _category_ to which some objects belong and some don't -- e.g. `/foo/` defines a category to which strings containing the substring "foo" belong and all other objects do not belong.  Not sure if that's a better or worse way to describe it but it's kinda nice to not use the "match" terminology for everything.

File: source/blog/2015-05-08-rspec-compsable-matchers.md [hunk 20]
```diff
@@ -0,0 +1,216 @@
...
+implements the `===` declares whether or not it "matches" the other object.
+For example, `Array === []` returns `true`. This is because an empty array is a
+subclass of the array class. `Array === {}` returns false because a hash is not
```
Comment (myronmarston) [inline source/blog/2015-05-08-rspec-compsable-matchers.md]: Shouldn't this be:

> This is because an empty array is an instance of the `Array` class.

...rather than:

> This is because an empty array is a subclass of the array class.

Given that every object has a singleton class (which is a one-off subclass of the instantiated class) the way you've phrased it isn't entirely inaccurate, but it's not how ruby programmers usually explain it.

File: source/blog/2015-05-08-rspec-compsable-matchers.md [hunk 21]
```diff
@@ -0,0 +1,216 @@
...
+a subclass of array. For regexes, `===` returns true if the passed string
```
Comment (myronmarston) [inline source/blog/2015-05-08-rspec-compsable-matchers.md]: Ditto here -- I think "because a hash is not an instance of `Array`" works better.

File: source/blog/2015-05-08-rspec-compsable-matchers.md [hunk 22]
```diff
@@ -0,0 +1,216 @@
...
+play around with it a little before moving on.
+
+## RSpec matchers
+
+In an RSpec expectation expression like `expect(object_1).to eq(object_2)`
+there are two primary objects at play, the **target** and the **matcher**. The
+**target** object is created by "expect" and basically wraps the passed object
+to it in a class called `ExpectationTarget`. The primary methods on the
+`ExpectationTarget` that users see are `to` and `not_to`/`to_not`. Let's focus
+on `to`.
```
Comment (myronmarston) [inline source/blog/2015-05-08-rspec-compsable-matchers.md]: Do you think that the details of `ExpectationTarget` is important for readers to understand composable matchers?

Comment (myronmarston) [inline source/blog/2015-05-08-rspec-compsable-matchers.md]: (I'm kinda thinking it's not, and it might be better to trim or cut this section).

File: source/blog/2015-05-08-rspec-compsable-matchers.md [hunk 23]
```diff
@@ -0,0 +1,216 @@
...
+passed matcher. It's possible to see this happening quite easily in an IRB session.
+
+```
+>> require "rspec/expectations"
+=> true
+>> o = Object.new
+=> #<Object:0x007fa20d221490>
+>> o.extend(RSpec::Matchers)
+=> #<Object:0x007fa20d221490>
+>> o.expect(3).to Object.new
```
Comment (myronmarston) [inline source/blog/2015-05-08-rspec-compsable-matchers.md]: I think it's confusing to show an expectation in this form, because all normal usage of expectations use bare `expect` rather than `some_object.expect`.  Rather than doing `o = Object.new; o.extend(RSpec::Matchers); o.expect(...)` you can do `include RSpec::Matchers; expect(...)`, which mirrors how users use the API.

File: source/blog/2015-05-08-rspec-compsable-matchers.md [hunk 24]
```diff
@@ -0,0 +1,216 @@
...
+>>
+```
+
+In this session, we extend the RSpec DSL on to a blank object and then call
+`o.expect(3).to Object.new`. This blows up because our new object does not implement
+the `matches?` method.
+
+In RSpec expectations, most matchers inherit from a class called `BaseMatcher`
+which causes `matches?` to delegate to a method called `match`. Some matchers have
+a very simple definition of `match`. For example, here's the definition of `match`
```
Comment (myronmarston) [inline source/blog/2015-05-08-rspec-compsable-matchers.md]: The fact that RSpec's built-in matchers subclass a base class which defines `matches?` to delegate to `match` doesn't seem important here.  `BaseMatcher` is definitely an implementation detail and is not part of RSpec's public API or semantics.  You also risk confusing the reader with the mention of `matches?` and `match` -- the fact that we've chosen to reduce duplication using that implementation mechanism has nothing to do with how composable matchers works but its presence here in the blog post might suggest that it does.  It's also confusing that the `matches?` protocol takes one arg (`actual`) but `match` takes two.

Instead, I'd recommend just talking about the matcher protocol (`matches?`) without reference to any extra layers in the built-in matchers such as `BaseMatcher` or `match`.  If you want to show what the definition is for `eq`, you could say something like;

> For example, here's the `matches?` logic for the `eq` matcher:
>
> ``` ruby
> def matches?(actual)
>   actual == expected
> end
> ```

File: source/blog/2015-05-08-rspec-compsable-matchers.md [hunk 25]
```diff
@@ -0,0 +1,216 @@
...
+for the `eq` matcher:
+
+```ruby
+def match(expected, actual)
+  actual == expected
+end
+```
+
+Some matchers, however, have a more complicated match protocol. This more complicated
+protocol is called the **composable** matcher protocol, which is what we'll discuss next.
```
Comment (myronmarston) [inline source/blog/2015-05-08-rspec-compsable-matchers.md]: I think you're overloading the term "protocol" here.  Our docs and other publications talk about the "matcher protocol" , and matchers which support composability implement that protocol.  As the author of the composability matcher support in RSpec 3, I'm not even sure what "the composable matcher protocol" refers to.

File: source/blog/2015-05-08-rspec-compsable-matchers.md [hunk 26]
```diff
@@ -0,0 +1,216 @@
...
+
+## Composable RSpec matchers
+
+A composable RSpec matcher is one which expects other RSpec matchers for some/all of it's
```
Comment (myronmarston) [inline source/blog/2015-05-08-rspec-compsable-matchers.md]: In a matcher like `start_with(a_value > 10)`, do you consider the composable matcher to be `start_with` or `a_value > 10`?  They are composed together but they have very different roles.  I think we need to differentiate them.

In the book we're using the term "primitive matcher" and "higher-order matcher":

* A primitive matcher is one that can't be broken down any further, but that can itself be used to built-up a larger composed matcher expression.  (`eq`, `be_predicate`, `be > x`, etc are all primitives).
* A higher order matcher is one that is capable of receiving another matcher as an argument and applying it in some fashion to its matching logic. `include`, `start_with`, `output`, etc are higher-order matchers.  Higher-order matchers can themselves to passed as arguments to other higher-order matchers.

I also think that "expects other RSpec matchers for some/all of its arguments" is confusing (and wrong, frankly). These matchers are capable of receiving other matchers as arguments, but they don't expect them -- they work just fine with non-matcher values.  In addition, the entire point of using `===` is so that they aren't even aware of what kind of object they are dealing with -- matcher or otherwise.  They can just use `===` as a universal protocol, which means regexes, ranges, classes, and matchers (at least those that define `===` properly) all work fine as arguments, and they can be blissfully unaware of what type of object they have been given.

File: source/blog/2015-05-08-rspec-compsable-matchers.md [hunk 27]
```diff
@@ -0,0 +1,216 @@
...
+```ruby
+def match(expected, actual)
+  return true if values_match?(expected, actual)
+  return false unless can_safely_call_match?(expected, actual)
+  actual.match(expected)
+end
+```
+
+The key important difference is that call to `values_match?` which is the core implementation
+of the new matching protocol in RSpec. The list of matchers which now reference the values_match?
```
Comment (myronmarston) [inline source/blog/2015-05-08-rspec-compsable-matchers.md]: I don't consider `values_match?` to be part of any kind of new protocol.  `===` is the only additional protocol at work for composablity.  `values_match?` is just a helper method that _leverages_ the `===` protocol but isn't itself a protocol.  There's only one implementation of `values_match?` and it does something concrete...it's not an abstract interface.

File: source/blog/2015-05-08-rspec-compsable-matchers.md [hunk 28]
```diff
@@ -0,0 +1,216 @@
...
+method are (as of RSpec expectations 3.3.0 development, not including mocks):
+
+* `change`
+* `contain_exactly`
+* `have_attributes`
+* `include`
+* `match`
+* `output`
+* `raise_error`
+* `start_or_end_with`
```
Comment (myronmarston) [inline source/blog/2015-05-08-rspec-compsable-matchers.md]: It's `start_with` and `end_with` as two separate matchers.

File: source/blog/2015-05-08-rspec-compsable-matchers.md [hunk 29]
```diff
@@ -0,0 +1,216 @@
...
+* `throw_symbol`
+* `yield`
```
Comment (myronmarston) [inline source/blog/2015-05-08-rspec-compsable-matchers.md]: There are 4 different `yield` matchers:

* `yield_control`, `yield_with_args`, `yield_with_no_args`, `yield_successive_args`.

File: source/blog/2015-05-08-rspec-compsable-matchers.md [hunk 30]
```diff
@@ -0,0 +1,216 @@
...
+  return true if values_match?(expected, actual)
+  return false unless can_safely_call_match?(expected, actual)
+  actual.match(expected)
+end
+```
+
+The key important difference is that call to `values_match?` which is the core implementation
+of the new matching protocol in RSpec. The list of matchers which now reference the values_match?
+method are (as of RSpec expectations 3.3.0 development, not including mocks):
+
```
Comment (myronmarston) [inline source/blog/2015-05-08-rspec-compsable-matchers.md]: If you choose to use the "primitive matchers" vs "higher-order matchers" terminology I've proposed above, this list would be the list of higher-order matchers.

File: source/blog/2015-05-08-rspec-compsable-matchers.md [hunk 31]
```diff
@@ -0,0 +1,216 @@
...
+  end
+end
+```
+
+The first thing to note is the special casing for Hashes and Arrays. Basically,
+what those methods do is recursively apply `values_match?` to each of the
+elements in the various collection types as you'd expect. Arrays by scanning
+elements and hashes by matching keys and values.
+
+The next thing the method does is return `true` if expected === actual. An easy
```
Comment (myronmarston) [inline source/blog/2015-05-08-rspec-compsable-matchers.md]: The next thing is `return true if expected == actual`, not `expected === actual`.

The `==` check is there because `/foo/ === /foo/` returns false as does `MyClass === MyClass` -- so we have to check normal equality as well and not rely solely on `===`.

File: source/blog/2015-05-08-rspec-compsable-matchers.md [hunk 32]
```diff
@@ -0,0 +1,216 @@
...
+
+```rspec
+def ===(value)
+  matches?(value)
+end
+```
+
+This means that `expected` in the expression `expected === actual` can be another RSpec matcher,
+a regex, or any other object that responds to `===`. This means, for example, that one can trivially
+match a lambda by doing `expect(obj).to match(lambda { |actual| ... })`. Given that all RSpec matchers
```
Comment (myronmarston) [inline source/blog/2015-05-08-rspec-compsable-matchers.md]: Unfortunately, not all RSpec matchers implement `===`.  All _built in_ matchers do, and all matches defined using the DSL do...but custom matchers defined as stand-alone classes do not unless the user has defined `===` or included `Composable`.

**Comment by myronmarston** [comment]
It'd probably be good to have a link to http://rspec.info/blog/2014/01/new-in-rspec-3-composable-matchers/ somewhere in your post, as that has lots of examples and motivation for the feature.

I've also got a suggestion for an alternate way to explain how composability works that I think readers will care much more about.  I think a key question RSpec users (particularly extension authors) have is, "how do I make my custom matchers composable like the built-in ones?".  Knowing what RSpec's bulit-in matchers do doesn't help the user know what they should do.  For example, you mention the `Composable` module, but you don't ever mention that it is intended for inclusion in custom matchers (and automatically included in custom matchers defined via the DSL).  Here's how I'm thinking you could demonstrate the concepts by showing how to make custom matchers composable:

* Q: I've got a custom matcher that I want to pass as an argument to one of RSpec's built in matchers.  How do I do that?
* A: we use the universal `===` protocol, so just `define `===` as an alias for `matches?` on your custom matcher.
* Actually, if you include `RSpec::Matchers::Composable` in your matcher class, it does that for you!  (And the DSL includes that automatically, so DSL-defined custom matchers already work this way!)
* The `Composable` module also defines `and` and `or` so you can use your custom matcher in compound matcher expressions, too.
* Q: the normal "verb" form of my matcher reads oddly when passed as an argument (and the failure message is really awkward grammatically)...what can I do about this?
* A: Use `RSpec::Matchers.alias_matcher`!  It reads better and provides a more readable failure message.
* Q: I want to define my own higher-order matcher that accepts other matchers as arguments.  How do I do that?
* A: The `Composable` module we defined before provides a helper method for this purpose: `values_match?`, which takes care of the matching for you, and understands nested array/hash data structures as well.
* Q: I've got my custom higher-order matcher working, but the failure messages don't use the description of the matcher argument like RSpec's built-in matchers do.  How do I fix that?
* A: The `Composable` module provides `description_of` and `surface_descriptions_in` to build good failure messages.

A sequence like this would explore all the concepts while also showing how the user can use those concepts in their own matchers.  It also avoids potential confusion around extra implementation details that your current blog post has.  For example, you show the logic for the `match` matcher, which includes a call to `can_safely_call_match?` without showing what that method is or what it does.  I think it's potentially confusing and doing examples with a custom matcher may help avoid that issue.

**Comment by myronmarston** [comment]
BTW, if you do decide to go with something like the alternate way I've sketched out here, don't feel like you have to follow that outline or even cover all the same points.  It was just meant to be an example of how you could present the information as a solution to the a problem a reader may have with a custom matcher, which I think will be more interesting than just showing how RSpec works internally.

**Comment by JonRowe** [comment]
Do you want someone else to finish this off @samphippen ?

**Comment by samphippen** [comment]
@JonRowe I was just cleaning up my PRs, but yeah if you've got some time (I don't). The other thing is, this post was really meant to help explain it to a friend and I sent it to him and that sort of served it's purpose. If you think this is close, totally feel free to finish it :)

**Comment by JonRowe** [comment]
I took a stab at making the changes to this, care to re-review @xaviershay / @myronmarston ?

File: source/blog/2015-09-22-rspec-compsable-matchers.md [hunk 33]
```diff
@@ -0,0 +1,218 @@
...
+that implement this protocol include:
+
+* all class objects
+* lambdas
+* regexes
+* ranges
+
+Conceptually, this protocol represents a "match", but can be thought of as
+categorising instances of objects. That is, the object that implements the
+`===` declares whether or not it belongs in the same category as the other
```
Comment (myronmarston) [inline source/blog/2015-09-22-rspec-compsable-matchers.md]: > the object that implements the `===` declares...

This should be "the object that implements the `===` method declares..." or "the object that implements `===` declares..."

Also, IMO, this explanation is inverted.  The object that implements `===` defines the category to which the passed object may (or may not) belong, but as written, "declares whether or not it belongs in the same category as the other object" reads like the inverse.

File: source/blog/2015-09-22-rspec-compsable-matchers.md [hunk 34]
```diff
@@ -0,0 +1,218 @@
...
+The new RSpec matchers make heavy use of `===` internally, if you'd like to get
+more of an understanding of `===`, I'd heavily suggest you pop open IRB and
+play around with it a little before moving on.
+
+## RSpec matchers
+
+In an RSpec expectation expression like `expect(object_1).to eq(object_2)`
+there are two primary objects at play, the **target** and the **matcher**. The
+**target** object is created by "expect" and exposes `to` (and `to_not`) which
+takes the matcher is is responsible for invoking `matches?` on it.
```
Comment (myronmarston) [inline source/blog/2015-09-22-rspec-compsable-matchers.md]: You have "is is" and this sentence is generally hard to read.

File: source/blog/2015-09-22-rspec-compsable-matchers.md [hunk 35]
```diff
@@ -0,0 +1,218 @@
...
+
+```ruby
+def matches?(actual)
+  actual == expected
+end
+```
+
+## Composable RSpec matchers
+
+A composable RSpec matcher is one which expects other RSpec matchers for it's
```
Comment (myronmarston) [inline source/blog/2015-09-22-rspec-compsable-matchers.md]: I think this sentence would be more accurate if you say "accepts other RSpec matchers" instead of "expects other RSpec matchers". "Expects" suggests that you can't pass non-matchers as arguments.

File: source/blog/2015-09-22-rspec-compsable-matchers.md [hunk 36]
```diff
@@ -0,0 +1,218 @@
...
+arguments. These are higher order matchers (as opposed to more primitive
+matchers such as `be_predicate` or `be > x`) such as `include`, `start_with`
```
Comment (myronmarston) [inline source/blog/2015-09-22-rspec-compsable-matchers.md]: This reads pretty awkwardly, particularly the "(as opposed...such as...) such as..." bit.

File: source/blog/2015-09-22-rspec-compsable-matchers.md [hunk 37]
```diff
@@ -0,0 +1,218 @@
...
+    # ArgumentError for `expected === actual`.
+    false
+  end
+end
+```
+
+The first thing to note is the special casing for Hashes and Arrays. Basically,
+what those methods do is recursively apply `values_match?` to each of the
+elements in the various collection types as you'd expect. Arrays by scanning
+elements and hashes by matching keys and values.
```
Comment (myronmarston) [inline source/blog/2015-09-22-rspec-compsable-matchers.md]: > Arrays by scanning elements and hashes by matching keys and values.

This sentence has no verb.  If you want to use this phrasing, I think it reads OK if you replace `.` with `--`, eg.g.:

> ...in the various collection types as you'd expect--arrays by scanning elements and hashes by matching keys and values.

File: source/blog/2015-09-22-rspec-compsable-matchers.md [hunk 38]
```diff
@@ -0,0 +1,218 @@
...
+IRB session:
+
+```ruby
+>> expect(/foo/).to match(/foo/)
+=> true
+```
+
+What we've done here is constructed a both a target and match matcher with the value /foo/.
+When the match matcher hits `values_match?` it sees that /foo/ is not an array or a hash, and so
+checks `expected == actual`, which in this case is `/foo/ == /foo/`, which is true.
```
Comment (myronmarston) [inline source/blog/2015-09-22-rspec-compsable-matchers.md]: I found this paragraph (and the prior one) to be quite confusing, even though I authored the code in question.  Maybe we can come up with a simpler way to make this point?

File: source/blog/2015-09-22-rspec-compsable-matchers.md [hunk 39]
```diff
@@ -0,0 +1,218 @@
...
+```
+
+The reason we're able to pass the `a_hash_including` matcher inside the
+`include` matcher is that `values_match?` will call `===` on the inner
+`a_hash_including` matcher, which will invoke its `matches?` method, which
+will allow it to match the complex inner hash.
+
+##Conclusions
+
+The old RSpec match protocol used to be a lot easier to understand. Each of the
```
Comment (myronmarston) [inline source/blog/2015-09-22-rspec-compsable-matchers.md]: I don't think the protocol was any easier to understand before.  The RSpec matcher protocol hasn't significantly changed.  It's just the internal implementation of many of our matchers has changed to delegate to this new fuzzy matching logic via `values_match?`.

**Comment by myronmarston** [comment]
Thanks, @JonRowe.  I left some more feedback.  I don't think it's quite in a publishable state yet.  Up to you if you want to keep working on it.

**Comment by JonRowe** [comment]
I've taken your suggestions into consideration, want to try another read through?

**Comment by JonRowe** [comment]
Ping @myronmarston :)

**Comment by JonRowe** [comment]
Hey @myronmarston did you ever have a chance to re-read this?

File: source/blog/2015-10-31-rspec-compsable-matchers.md [hunk 40]
```diff
@@ -0,0 +1,220 @@
...
+
+The `===` method is used in Ruby case statements to decide if a branch should
+be taken. For example
+
+```Ruby
+case 3
+when (1...18)
+  puts "hello"
+end
+```
```
Comment (myronmarston) [inline source/blog/2015-10-31-rspec-compsable-matchers.md]: This isn't rendering right:

screen shot 2016-01-19 at 11 33 19 pm

File: source/blog/2015-10-31-rspec-compsable-matchers.md [hunk 41]
```diff
@@ -0,0 +1,220 @@
...
+>> expect(3).to Object.new
+NoMethodError: undefined method `matches?' for #<Object:0x007fa20d1f4e68>
+    from /Users/sam/.gem/ruby/2.1.5/gems/rspec-expectations-3.2.1/lib/rspec/expectations/handler.rb:50:in `block in handle_matcher'
+    from /Users/sam/.gem/ruby/2.1.5/gems/rspec-expectations-3.2.1/lib/rspec/expectations/handler.rb:27:in `with_matcher'
+    from /Users/sam/.gem/ruby/2.1.5/gems/rspec-expectations-3.2.1/lib/rspec/expectations/handler.rb:48:in `handle_matcher'
+    from /Users/sam/.gem/ruby/2.1.5/gems/rspec-expectations-3.2.1/lib/rspec/expectations/expectation_target.rb:54:in `to'
+    from (irb):4
+    from /Users/sam/.rubies/ruby-2.1.5/bin/irb:11:in `<main>'
+>>
+```
```
Comment (myronmarston) [inline source/blog/2015-10-31-rspec-compsable-matchers.md]: This is really not rendering right:

screen shot 2016-01-19 at 11 33 59 pm

Comment (myronmarston) [inline source/blog/2015-10-31-rspec-compsable-matchers.md]: Apart from the rendering, the stack trace is pretty noisy as well and can probably be truncated.

File: source/blog/2015-10-31-rspec-compsable-matchers.md [hunk 42]
```diff
@@ -0,0 +1,220 @@
...
+
+In RSpec expectations matchers implement `matches?` to provide assertions for
+your test suite, some of these are complex but they can also be quite simple.
+For example, here's the logic of `matches?` for the `eq` matcher:
+
+```ruby
+def matches?(actual)
+  actual == expected
+end
+```
```
Comment (myronmarston) [inline source/blog/2015-10-31-rspec-compsable-matchers.md]: Again, not rendering right:

screen shot 2016-01-19 at 11 36 02 pm

File: source/blog/2015-10-31-rspec-compsable-matchers.md [hunk 43]
```diff
@@ -0,0 +1,220 @@
...
+
+The new RSpec matchers make heavy use of `===` internally, if you'd like to get
+more of an understanding of `===`, I'd heavily suggest you pop open IRB and
+play around with it a little before moving on.
+
+## RSpec matchers
+
+In an RSpec expectation expression like `expect(object_1).to eq(object_2)`
+there are two primary objects at play, the **target** and the **matcher**. The
+**target** object is created by "expect" and exposes `to` (and `to_not`) which
```
Comment (myronmarston) [inline source/blog/2015-10-31-rspec-compsable-matchers.md]: Any reason you put expect in quotes instead of backticks?  The backticks help make it clear that it's a method since it renders mono spaced.

File: source/blog/2015-10-31-rspec-compsable-matchers.md [hunk 44]
```diff
@@ -0,0 +1,220 @@
...
+    from (irb):4
+    from /Users/sam/.rubies/ruby-2.1.5/bin/irb:11:in `<main>'
+>>
+```
+
+In this session, we extend the RSpec DSL into the main object and then call
+`expect(3).to Object.new`. This blows up because our new object does not
+implement the `matches?` method.
+
+In RSpec expectations matchers implement `matches?` to provide assertions for
```
Comment (myronmarston) [inline source/blog/2015-10-31-rspec-compsable-matchers.md]: s/RSpec expectations/rspec-expectations/ (that makes it clear it's a single term that's a gem name).

Also, a comma between "expectations" and "matchers" seems appropriate.

File: source/blog/2015-10-31-rspec-compsable-matchers.md [hunk 45]
```diff
@@ -0,0 +1,220 @@
...
+place to start.
+
+In RSpec 2.xx the match matcher was mostly used to match regexes, and its
+`matches?` definition looked thus:
+
+```ruby
+def matches?(expected, actual)
+  actual.match expected
+end
+```
```
Comment (myronmarston) [inline source/blog/2015-10-31-rspec-compsable-matchers.md]: This code snippet also doesn't render correctly.  (I'll stop commenting on them now -- assume it's that way for all of them).

File: source/blog/2015-10-31-rspec-compsable-matchers.md [hunk 46]
```diff
@@ -0,0 +1,220 @@
...
+
+The first thing to note is the special casing for Hashes and Arrays. Basically,
+what those methods do is recursively apply `values_match?` to each of the
+elements in the various collection types as you'd expect, so matching Arrays by
+scanning elements and hashes by matching keys and values.
+
+We next check normal equality `==` so that if `expected == actual` we return
+`true`, we do this because objects of the same type, e.g. `/foo/ === /foo/`
+ir `MyClass === MyClass`, return `false` so we have to check normal equality
```
Comment (myronmarston) [inline source/blog/2015-10-31-rspec-compsable-matchers.md]: ir?

File: source/blog/2015-10-31-rspec-compsable-matchers.md [hunk 47]
```diff
@@ -0,0 +1,220 @@
...
+as well and not rely solely on `===`. We can demonstrate this is back in our
```
Comment (myronmarston) [inline source/blog/2015-10-31-rspec-compsable-matchers.md]: The first sentence of this paragraph reads like a run-on sentence.

**Comment by myronmarston** [comment]
Sorry for the delay.  It's been hard for me to feel motivated to spend time on this.

This still feels pretty rough, honestly, and not up to the level of polish of what we publish.

Writing is hard :(.
